### PR TITLE
Add LimitOrderHook

### PR DESCRIPTION
**Pull request by luiz-lvj**
Solves #11 

Adds LimitOrderHook to `src/general` folder.

#### PR Checklist

- [x] Tests
- [x] Documentation
- [ ] Changeset

### CODE REVIEW

File: src/general/LimitOrderHook.sol [hunk 1]
```diff
@@ -0,0 +1,797 @@
...
+import {Currency} from "v4-core/src/types/Currency.sol";
+import {SafeCast} from "v4-core/src/libraries/SafeCast.sol";
+import {PoolId} from "v4-core/src/types/PoolId.sol";
+import {CurrencySettler} from "src/utils/CurrencySettler.sol";
+import {FullMath} from "v4-core/src/libraries/FullMath.sol";
+import {BalanceDelta, toBalanceDelta} from "v4-core/src/types/BalanceDelta.sol";
+import {IUnlockCallback} from "v4-core/src/interfaces/callback/IUnlockCallback.sol";
+import {StateLibrary} from "v4-core/src/libraries/StateLibrary.sol";
+import {IERC20Minimal} from "v4-core/src/interfaces/external/IERC20Minimal.sol";
+import {console} from "forge-std/console.sol";
```
Comment (arr00) [inline src/general/LimitOrderHook.sol:20]: .

File: src/general/LimitOrderHook.sol [hunk 2]
```diff
@@ -0,0 +1,797 @@
...
+
+import {BaseHook} from "src/base/BaseHook.sol";
+import {Hooks} from "v4-core/src/libraries/Hooks.sol";
+import {PoolKey} from "v4-core/src/types/PoolKey.sol";
+import {IPoolManager} from "v4-core/src/interfaces/IPoolManager.sol";
+import {BeforeSwapDelta, BeforeSwapDeltaLibrary, toBeforeSwapDelta} from "v4-core/src/types/BeforeSwapDelta.sol";
+import {Currency} from "v4-core/src/types/Currency.sol";
+import {SafeCast} from "v4-core/src/libraries/SafeCast.sol";
+import {PoolId} from "v4-core/src/types/PoolId.sol";
+import {CurrencySettler} from "src/utils/CurrencySettler.sol";
```
Comment (arr00) [inline src/general/LimitOrderHook.sol:14]: lets keep imports grouped by source

**Comment by arr00** [comment]
Keep in mind that our documentation generator only reads dev tags. That's why we generally don't use `@notice` or `@param` tags.

File: src/general/LimitOrderHook.sol [hunk 3]
```diff
@@ -0,0 +1,797 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Uniswap Hooks (last updated v0.1.0) (src/general/LimitOrderHook.sol)
```
Comment (arr00) [inline src/general/LimitOrderHook.sol:2]: Current release is 1.0.0?

Comment (luiz-lvj) [reply src/general/LimitOrderHook.sol:2]: Yes, fixing that

File: src/general/LimitOrderHook.sol [hunk 4]
```diff
@@ -0,0 +1,797 @@
...
+    mapping(Epoch => EpochInfo) public epochInfos;
+
+    /**
+     * @dev Zero liquidity was attempted to be added or removed.
+     */
+    error ZeroLiquidity();
+
+    /* 
+     * @dev Limit order was placed in range
+    */
```
Comment (arr00) [inline src/general/LimitOrderHook.sol:178]:
```suggestion
    /** 
     * @dev Limit order was placed in range
     */
```
Or could also just do /// since its a single line comment.

File: src/general/LimitOrderHook.sol [hunk 5]
```diff
@@ -0,0 +1,797 @@
...
+            (, BalanceDelta feesAccrued) = poolManager.modifyLiquidity(
+                cancelData.key,
+                IPoolManager.ModifyLiquidityParams({
+                    tickLower: cancelData.tickLower,
+                    tickUpper: tickUpper,
+                    liquidityDelta: 0,
+                    salt: 0
+                }),
+                ZERO_BYTES
+            );
```
Comment (saucepoint) [inline src/general/LimitOrderHook.sol:548]: can this be optimized by collecting fees when cancelling an order? i.e. only doing 1x `.modifyLiquidity` call?

```solidity
(BalanceDelta cancelDelta, BalanceDelta feesAccrued) = poolManager.modifyLiquidity(liquidityDelta=cancelDelta.liquidityDelta)

if (feesAccrued.amount0()) poolManager.mint(...);
// ...

BalanceDelta principalDelta = cancelDelta - feesAccrued;
if (principalDelta.amount0()) poolManager.take(recipient);
```

File: src/general/LimitOrderHook.sol [hunk 6]
```diff
@@ -0,0 +1,797 @@
...
+
+    /**
+     * @dev Get the tick for a given pool.
+     *
+     * @param poolId The pool id.
+     * @return tick The tick.
+     */
+    function getTick(PoolId poolId) private view returns (int24 tick) {
+        (, tick,,) = poolManager.getSlot0(poolId);
+    }
```
Comment (saucepoint) [inline src/general/LimitOrderHook.sol:772]: this is unsafe in very obscure cases -- cetora found it in the v4 formal verification, and we refer to it as M02 behavior

for zeroForOne swaps, which moves `slot0.sqrtPriceX96` to exactly a `tickLower` of an existing position, `slot0.tick` is decremented by one

https://github.com/Uniswap/v4-core/pull/930/files#diff-d198b8552afdb46e50c8452630825f118160ea6f2da94677d8e6d48d7ceeac95R1231-R1243

---

The safest way to obtain a tick is:

```solidity
(uint160 sqrtPriceX96,,,) = poolManager.getSlot0(poolId);
tick = TickMath.getTickAtSqrtPrice(sqrtPriceX96);
```

File: src/general/LimitOrderHook.sol [hunk 7]
```diff
@@ -0,0 +1,797 @@
...
+
+    /**
+     * @dev Set the last tick lower.
+     *
+     * @param poolId The pool id.
+     * @param tickLower The tick lower.
+     */
+    function setTickLowerLast(PoolId poolId, int24 tickLower) private {
+        tickLowerLasts[poolId] = tickLower;
+    }
```
Comment (saucepoint) [inline src/general/LimitOrderHook.sol:714]: small nit and personal preference -- i prefer in-lining the storage write as `tickLowerLasts[poolId] = tickLower`, especially since were only writing storage in two logic flows

**Review by saucepoint — COMMENTED** [review]
only small nits here -- impl looks pretty good. didnt look at tests, but i would make sure some boundary cases are tested, trying to close orders mid-tick, closing/withdrawing when spot tick is on tickLower or tickUpper, etc etc

also thoughts on renaming Epoch to OrderId?

File: src/general/LimitOrderHook.sol (hunk 5 above)
Comment (luiz-lvj) [reply src/general/LimitOrderHook.sol:548]: Yeah, it's better to have only one `modifyLiquidity`, updating that

**Review by saucepoint — APPROVED** [review]
lgtm

**Review by ernestognw — COMMENTED** [review]
Was just frontran by @saucepoint lol, feel free to address those in another PR. Pretty much nits and documentation suggestions

**Comment by luiz-lvj** [comment]
> Was just frontran by @saucepoint lol, feel free to address those in another PR. Pretty much nits and documentation suggestions

Haha thanks @ernestognw I'll address the issues in a new PR

File: src/general/LimitOrderHook.sol [hunk 8]
```diff
@@ -0,0 +1,792 @@
...
+ *
+ * _Available since v1.1.0_
+ */
+contract LimitOrderHook is BaseHook, IUnlockCallback {
+    using StateLibrary for IPoolManager;
+    using OrderIdLibrary for OrderId;
+    using CurrencySettler for Currency;
+
+    /**
+     * @notice The info for each order id.
```
[the remark this answers is not in the document]
Comment (luiz-lvj) [reply src/general/LimitOrderHook.sol:78]: Yeah, this is an interesting discussion. I believe it would be nice to render structs and enums

File: src/general/LimitOrderHook.sol [hunk 9]
```diff
@@ -0,0 +1,790 @@
...
+import {BalanceDelta, toBalanceDelta} from "v4-core/src/types/BalanceDelta.sol";
+import {IPoolManager} from "v4-core/src/interfaces/IPoolManager.sol";
+import {IUnlockCallback} from "v4-core/src/interfaces/callback/IUnlockCallback.sol";
+import {CurrencySettler} from "src/utils/CurrencySettler.sol";
+import {BaseHook} from "src/base/BaseHook.sol";
+
+/**
+ * @dev The order id type.
+ */
+type OrderId is uint232;
```
[the remark this answers is not in the document]
Comment (luiz-lvj) [reply src/general/LimitOrderHook.sol:24]: It makes sense, thanks for pointing that out. For now I'll keep this library inside the `LimitOrderHook.sol` file (but with the OrderId inside it), but if we start to have multiple utilities for orders I will out it.

**Comment by luiz-lvj** [comment]
@ernestognw I opened PR #58 to address the issues you raised. We can continue the conversation there if needed.
